plutus-tx/src/PlutusTx/Data/List.hs

```diff
@@ -0,0 +1,194 @@
+import PlutusTx.Builtins.Internal qualified as BI
+import PlutusTx.IsData.Class (FromData (..), ToData (..), UnsafeFromData (..))
+import PlutusTx.Lift (makeLift)
+import PlutusTx.Prelude hiding (any, filter, find, findIndices, foldMap, length, map, mapMaybe,
+                         pred)
+import Prettyprinter (Pretty (..))
+
+import Data.Semigroup qualified as Haskell
+import Prelude qualified as Haskell
+
+-- | A list type backed directly by 'Data'. It is meant to be used whenever fast
+-- encoding/decoding to/from 'Data' is needed.
+newtype List a = List (BuiltinList BuiltinData)
+  deriving stock (Haskell.Show, Haskell.Eq)
+
+instance Eq (List a) where
+    {-# INLINEABLE (==) #-}
+    List l == List l' =
+        case (B.uncons l, B.uncons l') of
+            (Just (h, t), Just (h', t')) -> h == h' && List t == List t'
+            (Nothing, Nothing)           -> True
+            _                            -> False
+
+instance ToData (List a) where
+    {-# INLINEABLE toBuiltinData #-}
+    toBuiltinData (List l) = BI.mkList l
+
+instance FromData (List a) where
+    {-# INLINEABLE fromBuiltinData #-}
+    fromBuiltinData = Just . List . BI.unsafeDataAsList
+
+instance UnsafeFromData (List a) where
+    {-# INLINEABLE unsafeFromBuiltinData #-}
+    unsafeFromBuiltinData = List . BI.unsafeDataAsList
+
+instance (UnsafeFromData a, Pretty a) => Pretty (List a) where
+    {-# INLINEABLE pretty #-}
+    pretty = pretty . toSOP
+
+append :: List a -> List a -> List a
+append (List l) (List l') = List (go l l')
+  where
+    go xs ys =
+        B.caseList'
+            ys
+            (\h t -> BI.mkCons h (go t ys))
+            xs
+
+instance Semigroup (List a) where
+    (<>) = append
+
+instance Monoid (List a) where
+    mempty = List B.mkNil
+
+instance Haskell.Semigroup  (List a) where
+    (<>) = append
+
+instance Haskell.Monoid (List a) where
+    mempty = List B.mkNil
+
+{-# INLINEABLE toSOP #-}
+toSOP :: (UnsafeFromData a) => List a -> [a]
+toSOP (List l) = go l
+  where
+    go = B.caseList' [] (\h t -> unsafeFromBuiltinData h : go t)
+
+{-# INLINEABLE fromSOP #-}
+fromSOP :: (ToData a) => [a] -> List a
+fromSOP = List . BI.unsafeDataAsList . B.mkList . fmap toBuiltinData
+
+{-# INLINEABLE find #-}
+find :: (UnsafeFromData a) => (a -> Bool) -> List a -> Maybe a
+find pred (List l) = go l
+  where
+    go =
+        B.caseList'
+            Nothing
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                in
+                    if pred h'
+                        then Just h'
+                        else go t
+            )
+
+-- TODO: return [] or List?
+{-# INLINEABLE findIndices #-}
+findIndices :: (UnsafeFromData a) => (a -> Bool) -> List a -> [Integer]
+findIndices pred (List l) = go 0 l
+  where
+    go i =
+        B.caseList'
+            []
+            (\h t ->
+                let h' = unsafeFromBuiltinData h
+                    indices = go (B.addInteger i 1) t
```

Comment:
Please use `addInteger 1 t`. Search for "Yoda conditions" on Slack for why that makes any sense at all.